From 5ad42e3860aeb24e5d8bf87a03a4787008aeac85 Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Wed, 8 Jan 2025 17:58:27 +0100
Subject: [PATCH 1431/1695] plat-stm32mp2: add stm32mp2_tamp driver

Implement stm32mp2_tamp driver to enable plateform internal tamper
based on the tamper id.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 534d73f4aca95a8d50fc3350a468a43e9dbfdb4f plat-stm32mp2: add stm32mp2_tamp driver]
Change-Id: Ia8e68926403e92b3651a1a15756d7764a83c360e
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/427743
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 .../arm/plat-stm32mp2/drivers/stm32mp2_tamp.c | 91 +++++++++++++++++++
 .../arm/plat-stm32mp2/drivers/stm32mp2_tamp.h | 13 +++
 core/arch/arm/plat-stm32mp2/drivers/sub.mk    |  1 +
 core/arch/arm/plat-stm32mp2/main.c            | 75 ---------------
 core/arch/arm/plat-stm32mp2/stm32_util.h      |  2 -
 core/drivers/stm32_tamp.c                     |  3 +
 6 files changed, 108 insertions(+), 77 deletions(-)
 create mode 100644 core/arch/arm/plat-stm32mp2/drivers/stm32mp2_tamp.c
 create mode 100644 core/arch/arm/plat-stm32mp2/drivers/stm32mp2_tamp.h

diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp2_tamp.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp2_tamp.c
new file mode 100644
index 000000000..e41b4683d
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp2_tamp.c
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2025, STMicroelectronics
+ */
+
+#include <drivers/stm32_tamp.h>
+#include <drivers/stm32mp2_tamp.h>
+#include <drivers/stm32mp25_pwr.h>
+#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23)
+#include <drivers/stm32mp25_rcc.h>
+#endif
+#if defined(CFG_STM32MP21)
+#include <drivers/stm32mp21_rcc.h>
+#endif
+#include <io.h>
+#include <stm32_util.h>
+
+/* Activate the SoC resources required by internal TAMPER */
+TEE_Result stm32_activate_internal_tamper(int id)
+{
+	TEE_Result res = TEE_ERROR_NOT_SUPPORTED;
+
+	switch (id) {
+	case INT_TAMP1: /* Backup domain (V08CAP) voltage monitoring */
+	case INT_TAMP2: /* Temperature monitoring */
+		stm32mp_pwr_monitoring_enable(PWR_MON_V08CAP_TEMP);
+		res = TEE_SUCCESS;
+		break;
+
+	case INT_TAMP3: /* LSE monitoring (LSECSS) */
+		if (io_read32(stm32_rcc_base() + RCC_BDCR) & RCC_BDCR_LSECSSON)
+			res = TEE_SUCCESS;
+		break;
+
+	case INT_TAMP4: /* HSE monitoring (CSS + over frequency detection) */
+		if (io_read32(stm32_rcc_base() + RCC_OCENSETR) &
+		    RCC_OCENSETR_HSECSSON)
+			res = TEE_SUCCESS;
+		break;
+
+	case INT_TAMP7:
+		if (IS_ENABLED(CFG_STM32MP21)) {
+			/* ADC2 (adc2_awd1) analog watchdog monitoring 1 */
+			res = TEE_SUCCESS;
+			break;
+		} else if (IS_ENABLED(CFG_STM32MP23) ||
+			   IS_ENABLED(CFG_STM32MP25)) {
+			/* VDDCORE monitoring under/over voltage */
+			stm32mp_pwr_monitoring_enable(PWR_MON_VCORE);
+			res = TEE_SUCCESS;
+			break;
+		}
+		break;
+
+	case INT_TAMP12:
+		if (IS_ENABLED(CFG_STM32MP21)) {
+			/* ADC2 (adc2_awd2) analog watchdog monitoring 2 */
+			res = TEE_SUCCESS;
+			break;
+		} else if (IS_ENABLED(CFG_STM32MP23) ||
+			   IS_ENABLED(CFG_STM32MP25)) {
+			/* VDDCPU (Cortex A35) monitoring under/over voltage */
+			stm32mp_pwr_monitoring_enable(PWR_MON_VCPU);
+			res = TEE_SUCCESS;
+			break;
+		}
+		break;
+
+	case INT_TAMP13:
+	case INT_TAMP16:
+		if (IS_ENABLED(CFG_STM32MP21))
+			res = TEE_SUCCESS;
+		break;
+
+	case INT_TAMP5:
+	case INT_TAMP6:
+	case INT_TAMP8:
+	case INT_TAMP9:
+	case INT_TAMP10:
+	case INT_TAMP11:
+	case INT_TAMP14:
+	case INT_TAMP15:
+		res = TEE_SUCCESS;
+		break;
+
+	default:
+		break;
+	}
+
+	return res;
+}
diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp2_tamp.h b/core/arch/arm/plat-stm32mp2/drivers/stm32mp2_tamp.h
new file mode 100644
index 000000000..1e596d3d8
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp2_tamp.h
@@ -0,0 +1,13 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2025, STMicroelectronics
+ */
+
+#ifndef __STM32MP2_TAMP_H
+#define __STM32MP2_TAMP_H
+
+#include <tee_api_types.h>
+
+TEE_Result stm32_activate_internal_tamper(int id);
+
+#endif /* __STM32MP2_TAMP_H */
diff --git a/core/arch/arm/plat-stm32mp2/drivers/sub.mk b/core/arch/arm/plat-stm32mp2/drivers/sub.mk
index 58a6af05d..bbc9400a9 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/sub.mk
+++ b/core/arch/arm/plat-stm32mp2/drivers/sub.mk
@@ -2,4 +2,5 @@ srcs-y += stm32mp25_syscfg.c
 srcs-$(CFG_STM32_PWR) += stm32mp25_pwr.c
 srcs-$(CFG_STM32_PWR_IRQ) += stm32mp25_pwr_irq.c
 srcs-$(CFG_STM32_PWR_REGUL) += stm32mp25_pwr_regul.c
+srcs-$(CFG_STM32_TAMP) += stm32mp2_tamp.c
 srcs-$(CFG_STPMIC2) += stm32mp2_pmic2.c
diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index 580628e83..ae7460bfe 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -311,81 +311,6 @@ void __noreturn do_reset(const char *str __maybe_unused)
 	panic();
 }
 
-/* Activate the SoC resources required by internal TAMPER */
-TEE_Result stm32_activate_internal_tamper(int id)
-{
-	TEE_Result res = TEE_ERROR_NOT_SUPPORTED;
-
-	switch (id) {
-	case INT_TAMP1: /* Backup domain (V08CAP) voltage monitoring */
-	case INT_TAMP2: /* Temperature monitoring */
-		stm32mp_pwr_monitoring_enable(PWR_MON_V08CAP_TEMP);
-		res = TEE_SUCCESS;
-		break;
-
-	case INT_TAMP3: /* LSE monitoring (LSECSS) */
-		if (io_read32(stm32_rcc_base() + RCC_BDCR) & RCC_BDCR_LSECSSON)
-			res = TEE_SUCCESS;
-		break;
-
-	case INT_TAMP4: /* HSE monitoring (CSS + over frequency detection) */
-		if (io_read32(stm32_rcc_base() + RCC_OCENSETR) &
-		    RCC_OCENSETR_HSECSSON)
-			res = TEE_SUCCESS;
-		break;
-
-	case INT_TAMP7:
-		if (IS_ENABLED(CFG_STM32MP21)) {
-			/* ADC2 (adc2_awd1) analog watchdog monitoring 1 */
-			res = TEE_SUCCESS;
-			break;
-		} else if (IS_ENABLED(CFG_STM32MP23) ||
-			   IS_ENABLED(CFG_STM32MP25)) {
-			/* VDDCORE monitoring under/over voltage */
-			stm32mp_pwr_monitoring_enable(PWR_MON_VCORE);
-			res = TEE_SUCCESS;
-			break;
-		}
-		break;
-
-	case INT_TAMP12:
-		if (IS_ENABLED(CFG_STM32MP21)) {
-			/* ADC2 (adc2_awd2) analog watchdog monitoring 2 */
-			res = TEE_SUCCESS;
-			break;
-		} else if (IS_ENABLED(CFG_STM32MP23) ||
-			   IS_ENABLED(CFG_STM32MP25)) {
-			/* VDDCPU (Cortex A35) monitoring under/over voltage */
-			stm32mp_pwr_monitoring_enable(PWR_MON_VCPU);
-			res = TEE_SUCCESS;
-			break;
-		}
-		break;
-
-	case INT_TAMP13:
-	case INT_TAMP16:
-		if (IS_ENABLED(CFG_STM32MP21))
-			res = TEE_SUCCESS;
-		break;
-
-	case INT_TAMP5:
-	case INT_TAMP6:
-	case INT_TAMP8:
-	case INT_TAMP9:
-	case INT_TAMP10:
-	case INT_TAMP11:
-	case INT_TAMP14:
-	case INT_TAMP15:
-		res = TEE_SUCCESS;
-		break;
-
-	default:
-		break;
-	}
-
-	return res;
-}
-
 #ifdef CFG_STM32_HSE_MONITORING
 /* pourcent rate of hse alarm */
 #define HSE_ALARM_PERCENT	110
diff --git a/core/arch/arm/plat-stm32mp2/stm32_util.h b/core/arch/arm/plat-stm32mp2/stm32_util.h
index cbacdd526..cd236415b 100644
--- a/core/arch/arm/plat-stm32mp2/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp2/stm32_util.h
@@ -26,8 +26,6 @@ static inline void stm32mp_register_gpioz_pin_count(size_t count __unused) { }
 
 void __noreturn do_reset(const char *str);
 
-TEE_Result stm32_activate_internal_tamper(int id);
-
 #ifdef CFG_STM32_CPU_OPP
 bool stm32mp_supports_cpu_opp(uint32_t opp_id);
 #endif /*CFG_STM32_CPU_OPP*/
diff --git a/core/drivers/stm32_tamp.c b/core/drivers/stm32_tamp.c
index 03a8362e6..927a8a871 100644
--- a/core/drivers/stm32_tamp.c
+++ b/core/drivers/stm32_tamp.c
@@ -25,6 +25,9 @@
 #if defined(CFG_STM32MP21)
 #include <drivers/stm32mp21_rcc.h>
 #endif
+#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23) || defined(CFG_STM32MP21)
+#include <drivers/stm32mp2_tamp.h>
+#endif /* defined(CFG_STM32MP25) || defined(CFG_STM32MP23) || defined(CFG_STM32MP21) */
 #include <drivers/stm32mp_dt_bindings.h>
 #include <io.h>
 #include <kernel/boot.h>
-- 
2.47.2

